### PR TITLE
Fixed feature bug and added embeddings to transducer.py

### PR DESCRIPTION
**Pull request by bonham79**
Fixes https://github.com/CUNY-CL/yoyodyne/issues/12 and https://github.com/CUNY-CL/yoyodyne/issues/61 . Lays groundwork for https://github.com/CUNY-CL/yoyodyne/issues/47 .

Fixes are as follows:
-- Addresses feature bug by adding a `features_map` to `IndexFeatures` that tracks encodings for features in dataset indexes. This can be utilized down the road when we expand feature encoding functionality.
-- This feature map encodes `special` encodings that parallels encoding for `source` and `target` strings. Doing so allows sharing of index locations across mappings. Please do not reorder the appending of `special.SPECIAL` in the `SymbolMap` class or things will break.
-- Adds a `feature_offset` to collator for when features and source strings are collated as a single string. This provides 'just in time' concatenation of the two to avoid further headaches when managing feature encodings at the dataloader level.
-- Updates `Collator` instantiation to utilize just a `dataset` object instead of multiple attributes from same object. 
-- Adds a `feature_embeddings` to `pointer_generator` to take advantage of direct referencing of `feature vocabulary size.` This will come in handy when we add in more feature customization.
--Updates `transducer` to use embeddings and brings its behavior more consistent with other models. Further enhancements can be made after we add more feature utility.

@Adamits  Would you make sure there was no breaking of `feature_invariant_transformer` functionality? I'm not up to date on that architecture so want to double check.

### CODE REVIEW

**Comment by kylebgorman** [comment]
Would prefer you to edit the tests to make stuff pass. Feel free to expand on them within reason.On Apr 10, 2023, at 6:43 PM, Travis Bartley ***@***.***> wrote:﻿
@kylebgorman Since I messed around with Collator some tests are gonna be broken. Want me to defer to instantiation that conforms to the tests or edit the tests to conform to new instantiation?

File: yoyodyne/models/pointer_generator.py
Comment (Adamits) [inline yoyodyne/models/pointer_generator.py:514]: Question: Did batch.features previously index the embeddings matrix incorrectly? Do you have any idea when this was introduced?

Comment (bonham79) [reply yoyodyne/models/pointer_generator.py:514]: I remember indexing behavior being correct prior to the `indexes` implementation. So it may have been introduced during that update.

File: yoyodyne/train.py
Comment (Adamits) [inline yoyodyne/train.py:313]: Is this only here for the feature invariant transformer now? I think we could make the same change to that module. The features_idx is only used to zero out positional embeddings for features, since their positional info is irrelevant.

File: yoyodyne/train.py
Comment (Adamits) [inline yoyodyne/train.py:293]: Should this perhaps be in the `Index`? Or does it need to be here since it depends on the model config (do we want to separate features or not)?

File: yoyodyne/models/transducer.py
Comment (Adamits) [inline yoyodyne/models/transducer.py:625]: Was this averaging in one of the publications for this model? If not, we should leave an option for the original behavior for reproducibility (I ask b/c it seems like features were handled differently before this PR).

File: yoyodyne/models/pointer_generator.py
Comment (bonham79) [reply yoyodyne/models/pointer_generator.py:514]: Ah, I think I know why this behavior wasn't affected. Since `pointer_generator` just uses one embedding and was instantiating from vocabulary, it was already the size of the conjoined `source + feature` space.  Since `batch.features` was still indexed in order with source vocabulary, it would have had the proper encodings.

File: yoyodyne/collators.py
Comment (Adamits) [inline yoyodyne/collators.py:95]: Does this get run if there are no features? What does `cat` do in that case?

**Review by Adamits — COMMENTED** [review]
I left a couple inline comments. They may not require any changes, but I want to be sure before I approve.

WRT feature invariant transformer, as long as the features_index is correctly computed, everything should work fine. 

I tried to make sure that I understand how everything works and it seems good to me, but there is certainly a chance that I missed something -- especially since I have not touched the indexing stuff in a long time (probably not since long before the `Indexes.py` stuff was added). Do we have a reliable way to test this right now?

As I understand this, we now index characters and features totally separately, and we combine them in the collator, by adding the source index length to all feature indices, for models that expect a single input string, and have a single embedding matrix? I think this seems good.

**Comment by bonham79** [comment]
>  Do we have a reliable way to test this right now?

I ran all models over SIGMORPHON 2021 Polish data set and all instances where feature was included I saw a performance increase (same seed). That's the only test I had atm though. Can try on Basque since I remember that dataset _only_ sees improvement if features are included.

> As I understand this, we now index characters and features totally separately, and we combine them in the collator, by adding the source index length to all feature indices, for models that expect a single input string, and have a single embedding matrix? I think this seems good.

That is correct. It seems an easier way to refer to features without needing to do indexing math and leaves the point of failure only to the `Collator` method.

File: tests/collator_test.py
Comment (kylebgorman) [inline tests/collator_test.py:38]: Index doesn't have a `pad_idx` or `source_vocab_size` in either your version or mine.  Not sure what's going on here.

**Comment by bonham79** [comment]
@kylebgorman @Adamits apparently my comments the first go around never went through, can you confirm if you received them this time?

**Comment by kylebgorman** [comment]
No recent comments from you are here on GitHub nor in my inbox.

**Comment by bonham79** [comment]
> No recent comments from you are here on GitHub nor in my inbox.

Oh that's annoying, all were pending because git thought I started a review...

**Comment by kylebgorman** [comment]
It won't let me comment on any of your self-review comments, but I can add that you can TODO the boilerplate stuff in the test like so:

```
# TODO: reduce boilerplate.
```

**Comment by bonham79** [comment]
> It won't let me comment on any of your self-review comments, but I can add that you can TODO the boilerplate stuff in the test like so:
> 
> ```
> # TODO: reduce boilerplate.
> ```

It actually gets fixed in my other PR. Should i add the TODO just for version history tracking or nah?

**Comment by kylebgorman** [comment]
In that case no need, trust you on that ;)

K

**Comment by bonham79** [comment]
> In that case no need, trust you on that ;) K

aight, think this is good to push then.

File: tests/collator_test.py
Comment (kylebgorman) [reply tests/collator_test.py:38]: Okay I'll come back to this later.

File: yoyodyne/datasets.py
Comment (kylebgorman) [inline yoyodyne/datasets.py:85]: Is this still needed? If so it should be `==` not `is`, I think

**Review by kylebgorman — APPROVED** [review]
This looks good to me now. Maybe add a TODO with the collator test thing. LGTM.

**Pull request merged**
